### DISCUSSION TITLE
Viewport box center hook attribute

### DISCUSSION THREAD

**Discussion by harsohailB** [xyflow/xyflow#1253] [New Features]
I think it would be useful to have a `useStoreState` attribute that keeps track of the center of the view box based on `width`, `height`, `transform`, and `zoom`. 

For my current use case, I needed this to render a new process in the center of wherever the user has moved the flow to. I achieved it with the following code, however, I am willing to work on this to make it more dynamic: 

```js
const currTransform = useStoreState((state) => state.transform);
const currViewportBox = useStoreState((state) => state.viewportBox);

const renderNewProcess = () => {
    // Variables
    const CUSTOM_NODE_WIDTH = 200;
    const AVG_CUSTOM_NODE_HEIGHT = 175;
    const [currTransformX, currTransformY, currZoom] = currTransform;
    const currZoomMultiplier = 1 / currZoom;

    // Calculations
    const centerX =
      -currTransformX * currZoomMultiplier +
      (currViewportBox.width * currZoomMultiplier) / 2;
    const centerY =
      -currTransformY * currZoomMultiplier +
      (currViewportBox.height * currZoomMultiplier) / 2;
    const nodeWidthCorrection = (CUSTOM_NODE_WIDTH / 2) * currZoomMultiplier;
    const nodeHeightCorrection =
      (AVG_CUSTOM_NODE_HEIGHT / 2) * currZoomMultiplier;

    // New process object
    const newProcess = {
      ...
      data: {
         ...
         position: {
           x: centerX - nodeWidthCorrection,
           y: centerY - nodeHeightCorrection,
         },
    };

    setElements((currElements) => currElements.concat(newProcess));
  };
```

**Comment by cubecleveland** [comment]
HI Can u give an example using this hook, im currently tryng to center a node in the middle of the container but i cant find the right way to do it

**Reply by harsohailB** [reply]
The calculation shown in the code above works to center a node in the middle of the react-flow canvas. I am awaiting validation from a dev on the react-flow repo to start building the hook.

**Reply by rianmartins** [reply]
Hi @harsohailB do you have this hook already implemented and available in the latest version of `reactflow` ?
